build.sbt

```diff
@@ -21,30 +22,50 @@ lazy val commonSettings = Seq(
   cancelable := true
 )
 
-lazy val bench = crossProject(JVMPlatform, NativePlatform)
+lazy val benchSettings = commonSettings ++ Seq(
+  buildInfoPackage := "rsc.bench",
+  buildInfoUsePackageAsPath := true,
+  buildInfoKeys := Seq[BuildInfoKey](
+    "sourceRoot" -> (baseDirectory in ThisBuild).value
+  )
+)
+
+lazy val benchJavac18 = project
+  .in(file("bench/javac18"))
+  .enablePlugins(BuildInfoPlugin)
+  .enablePlugins(JmhPlugin)
+  .settings(benchSettings)
+
+lazy val benchRsc = crossProject(JVMPlatform, NativePlatform)
   .crossType(CrossType.Full)
-  .in(file("bench"))
-  .dependsOn(rsc, tests)
+  .in(file("bench/rsc"))
+  .dependsOn(rsc)
+  .enablePlugins(BuildInfoPlugin)
   .enablePlugins(JmhPlugin)
+  .jvmSettings(benchCliRscNative("Typecheck"))
+  .settings(benchSettings)
+lazy val benchRscJVM = benchRsc.jvm
+lazy val benchRscNative = benchRsc.native
+
+lazy val benchScalac211 = project
+  .in(file("bench/scalac211"))
   .enablePlugins(BuildInfoPlugin)
-  .jvmSettings(
-    libraryDependencies += "org.scala-lang" % "scala-compiler" % versions.scala,
-    benchCliRscNative("Typecheck"),
-    benchCliRsc("Typecheck"),
-    benchCliScalac("Typecheck"),
-    benchCliScalac("Compile"),
-    benchCliJavac("Compile")
+  .enablePlugins(JmhPlugin)
+  .settings(
+    benchSettings,
+    scalaVersion := versions.scala211,
+    libraryDependencies += "org.scala-lang" % "scala-compiler" % scalaVersion.value
   )
+
+lazy val benchScalac212 = project
```

Comment:
@eed3si9n @dwijnand Can you provide some advice on deduplicating code in `benchScalac211` and `benchScalac212`? Files in these two projects are currently completely identical, which is clearly suboptimal, but I don't know how to make sbt do my bidding otherwise.

My goal is to be able to say `benchScalac211/jmh:run ...` or `benchScalac212/jmh:run ...` and then have sbt run the same benchmarks for either Scala 2.11 or Scala 2.12.

Here's what I tried so far:
  * One project in conjunction with `++`. This works, but I don't like the fact that users will have to know exact Scala versions (e.g. 2.12.4 vs simply 2.12) to run benchmarks. Even worse, if the user forgets to say `++`, they may inadvertently run benchmarks for an unexpected version Scala. I think this is a big deal, because benchmarking is monotonous enough to make humans accidentally miss things like that.
  * Two projects that are rooted at the same directory. This seems to be illegal.
  * Three projects (a common project that crosscompiles to Scala 2.11 and Scala 2.12, as well as two projects that depend on it). This leads to an error similar to the one described at https://stackoverflow.com/questions/27929272/how-to-have-sbt-subproject-with-multiple-scala-versions.

Reply:
It’s a bit hacky but I think you can use separate empty project directories and set them to have the same managed sources (or additional sources) to get around the fact that two projects can’t have the same directory.

Reply:
Interesting idea, thanks! Let's keep this in mind for the future.

Reply:
Sorry, with all the reasonable-scala repo activity I missed this notification was directly at me.

I too would go the way of adding a shared source directory to both projects.